neurom/core/population.py

```diff
@@ -27,42 +27,83 @@
 # SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 """Neuron Population Classes and Functions."""
+import logging
 
-from itertools import chain
+from morphio import MorphioError
+import neurom
+from neurom.exceptions import NeuroMError
+
+
+L = logging.getLogger(__name__)
 
 
 class Population:
     """Neuron Population Class.
 
-    Features:
-        - flattened collection of neurites.
-        - collection of somas, neurons.
-        - iterable-like iteration over neurons.
+    Offers an iterator over neurons within population, neurites of neurons, somas of neurons.
+    It does not store the loaded neuron in memory unless the neuron has been already passed
+    as loaded (instance of ``Neuron``).
     """
-    def __init__(self, neurons, name='Population'):
+    def __init__(self, files, name='Population', ignored_exceptions=()):
         """Construct a neuron population.
 
         Arguments:
-            neurons: iterable of neuron objects.
-            name: Optional name for this Population.
+            files (collections.abc.Sequence[str|Path|Neuron]): collection of neuron files or
+              paths to them
+            name (str): Optional name for this Population
+            ignored_exceptions (tuple): NeuroM and MorphIO exceptions that you want to ignore when
+              loading neurons.
         """
-        self.neurons = tuple(neurons)
-        self.somata = tuple(neu.soma for neu in neurons)
-        self.neurites = tuple(chain.from_iterable(neu.neurites for neu in neurons))
+        self._files = files
+        self._ignored_exceptions = ignored_exceptions
         self.name = name
 
+    @property
+    def neurons(self):
+        """Iterator to populations's somas."""
+        return (n for n in self)
+
+    @property
+    def somata(self):
+        """Iterator to populations's somas."""
+        return (n.soma for n in self)
+
+    @property
+    def neurites(self):
+        """Iterator to populations's neurites."""
+        return (neurite for n in self for neurite in n.neurites)
+
+    def _load_file(self, f):
+        if isinstance(f, neurom.core.neuron.Neuron):
+            return f
+        try:
+            return neurom.load_neuron(f)
+        except (NeuroMError, MorphioError) as e:
+            if isinstance(e, self._ignored_exceptions):
+                L.info('Ignoring exception "%s" for file %s', e, f.name)
+            else:
+                raise NeuroMError('`load_neurons` failed') from e
+        return None
+
     def __iter__(self):
         """Iterator to populations's neurons."""
-        return iter(self.neurons)
+        for f in self._files:
+            nrn = self._load_file(f)
+            if nrn is None:
+                continue
+            yield nrn
```

Comment:
Souldn't we use a cache mechanism here to avoid loading the neurons each time we request one? If I am correct, if we call `pop.neurons` then `pop.neurites`, all the files are loaded twice, which might be quite heavy.

Reply:
The whole point of this PR is to avoid loading into memory, so it can handle huge amounts of data. Solution to your concern will be to use parallelization. Also it is still possible to keep everything in memory as `Population` accepts a list of loaded neurons as well.

Reply:
I don't get your point on parallelization. Even if we parallelize the code to work on batches of neurons, each neuron file will still be loaded twice, isn't it? But if we want to avoid that it is indeed possible to load the neurons and then create the population from it, so I guess it's no big deal.

Reply:
Well, as an alternative a flag `cache=True/False` can be introduced. When `True`, all loaded morphologies will be kept in memory.

Reply:
> neuron file will still be loaded twice, isn't it?

Yes. A neuron will be loaded each time it is accessed within a separate iteration.

Reply:
> Well, as an alternative a flag `cache=True/False` can be introduced. When `True`, all loaded morphologies will be kept in memory.

Indeed this would be very handy.